flink/src/test/java/org/apache/iceberg/flink/TestChangeLogTable.java

```diff
@@ -0,0 +1,302 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.flink;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import org.apache.flink.types.Row;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.iceberg.BaseTable;
+import org.apache.iceberg.CatalogProperties;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.TableMetadata;
+import org.apache.iceberg.TableOperations;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.data.Record;
+import org.apache.iceberg.flink.source.BoundedTableFactory;
+import org.apache.iceberg.flink.source.ChangeLogTableTestBase;
+import org.apache.iceberg.relocated.com.google.common.base.Joiner;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.util.StructLikeSet;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+@RunWith(Parameterized.class)
+public class TestChangeLogTable extends ChangeLogTableTestBase {
```

Comment:
it seems there could be many combinations of changelog sequence. how do we know that we covered the meaningful combinations?

Reply:
Actually we cannot cover all the combinations in this unit test, but I think we could cover those critical cases in this class: 

1.  The first dimension is the primary key columns (PK), we choose to cover the three cases: 
    a) `id` as PK ;
    b) `data` as PK;
    c) `id,data` as PK;
The different primary key from different combinations will lead to different path to [generate the key of a row](https://github.com/apache/iceberg/blob/90225d6c9413016d611e2ce5eff37db1bc1b4fc5/core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java#L122).  We need to ensure that the data correctness  is not affected by the key combination.  

2. The second dimension is:  What's the correct results after applying INSERT, DELETE, UPDATE inside a same transaction.   The same key may have different version of INSERT, DELETE, UPDATE.  We need to ensure that the various insert/delete/update combinations within a given txn will not affect the correctness.

3.  The third dimension is: What's the correct results after applying INSERT/DELETE/UPDATE among different transaction.  We need to ensure that the various insert/delete/update combinations between different txn will not affect the correctness.

This's the reason that why do we choose to design the following test cases.

Reply:
thx. might be helpful to add those in comments to give readers some context on the unit test design